engine/src/main/java/org/hibernate/search/backend/impl/lucene/LuceneBackendResources.java

```diff
@@ -66,28 +87,42 @@ private LuceneBackendResources(LuceneBackendResources previous) {
 		this.workspace = previous.workspace;
 		this.maxQueueLength = previous.maxQueueLength;
 		this.asyncIndexingExecutor = previous.asyncIndexingExecutor;
```

Comment:
Shouldn't this be protected by locking `previous.executorStateReadLock`? Since the attribute is no longer `volatile`, that seems necessary... ?

Also, I wonder if the whole executor + readLock + writeLock stuff shouldn't be encapsulated in a separate object... What I find dangerous is that just after we copied the executor, one could call `shutdown` on a previous `LuceneBackendResources`, which would shut down the executor but would not clean up the `asyncIndexingExecutor` attribute on any new `LuceneBackendResources` that happened to share the same executor. And then we would have tasks being submitted but never handled.

Having a single, shared object hold the references to the locks would *and* to the executor would probably solve the issue. Then, when starting or shutting down the executor, we wouldn't change an attribute of `LuceneBackendResources`, but an attribute of that single, shared object that happens to be referenced from `LuceneBackendResources`, which is slightly different.

Or maybe there's another mechanism that guarantees that we'll never call a method on a previous version of the resources after having created a new version?